### PR TITLE
feat:Support BlockHashOrNumber with block_with_senders

### PR DESCRIPTION
**Pull request by supernovahs**
Ref #5302

### CODE REVIEW

**Comment by supernovahs** [comment]
Looks like my  nightly fmt is doing some weird formatting  😅

File: crates/storage/provider/src/providers/database/provider.rs [hunk 1]
```diff
@@ -1043,21 +1043,31 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
...
-        block_number: BlockNumber,
+        block_number: BlockHashOrNumber,
         transaction_kind: TransactionVariant,
     ) -> RethResult<Option<BlockWithSenders>> {
-        let Some(header) = self.header_by_number(block_number)? else { return Ok(None) };
+        let block_num: BlockNumber = match block_number {
+            BlockHashOrNumber::Hash(hash) => {
+                let res = match self.block_number(hash)? {
+                    Some(num) => Ok(num),
+                    None => Err("Block number not found for given hash"),
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:1053]: this should return the appropriate error

Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:1053]: this needs to return directly

File: crates/storage/provider/src/providers/database/provider.rs [hunk 2]
```diff
@@ -1043,21 +1043,31 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
...
+                };
+                res.unwrap()
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:1055]: redundant

**Review by mattsse — CHANGES_REQUESTED** [review]
yeah please use nightly formatting and undo `;`

File: crates/storage/provider/src/providers/database/provider.rs [hunk 3]
```diff
@@ -1043,21 +1043,31 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
...
+            }
+            BlockHashOrNumber::Number(num) => num,
+        };
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/database/provider.rs:1058]: can we make use of existing helper function here? https://github.com/paradigmxyz/reth/blob/4865e59684fbf1457125feec0c06a30405bf4b60/crates/storage/provider/src/traits/block_id.rs#L99-L105

@mattsse wdyt?

File: crates/storage/provider/src/providers/database/provider.rs [hunk 4]
```diff
@@ -1043,21 +1043,31 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
     /// will return None.
     fn block_with_senders(
         &self,
-        block_number: BlockNumber,
+        block_number: BlockHashOrNumber,
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/database/provider.rs:1046]: not a block number anymore
```suggestion
        block: BlockHashOrNumber,
```

File: crates/storage/provider/src/providers/database/provider.rs (hunk 3 above)
Comment (mattsse) [reply crates/storage/provider/src/providers/database/provider.rs:1058]: sgtm!

File: crates/storage/provider/src/providers/database/provider.rs [hunk 5]
```diff
@@ -1043,21 +1043,23 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
     /// will return None.
     fn block_with_senders(
         &self,
-        block_number: BlockNumber,
+        id: BlockHashOrNumber,
         transaction_kind: TransactionVariant,
     ) -> RethResult<Option<BlockWithSenders>> {
-        let Some(header) = self.header_by_number(block_number)? else { return Ok(None) };
+        let block_num = self.convert_hash_or_number(id).unwrap().unwrap();
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/database/provider.rs:1049]: please handle the error properly here

Comment (shekhirin) [inline crates/storage/provider/src/providers/database/provider.rs:1049]:
```suggestion
        let block_number = self.convert_hash_or_number(id).unwrap().unwrap();
```

**Review by shekhirin — DISMISSED** [review]
two small changes, then LGTM

File: crates/stages/src/stages/execution.rs [hunk 6]
```diff
@@ -147,7 +147,10 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
 
             // we need the block's transactions but we don't need the transaction hashes
             let block = provider
-                .block_with_senders(block_number, TransactionVariant::NoHash)?
+                .block_with_senders(
+                    BlockHashOrNumber::Number(block_number),
```
Comment (mattsse) [inline crates/stages/src/stages/execution.rs:151]: use .into()

File: crates/storage/provider/src/providers/database/provider.rs [hunk 7]
```diff
@@ -1043,9 +1043,14 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
...
     fn block_with_senders(
         &self,
-        block_number: BlockNumber,
+        id: BlockHashOrNumber,
         transaction_kind: TransactionVariant,
     ) -> RethResult<Option<BlockWithSenders>> {
+        let block_number = match self.convert_hash_or_number(id)? {
+            Some(number) => number,
+            None => return Ok(None),
+        };
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:1052]: let's also use let Some else return, see below

**Review by mattsse — CHANGES_REQUESTED** [review]
smol style nits

**Review by mattsse — APPROVED** [review]
lgtm, ty

**Pull request merged**